File: strings/levenshtein_distance.js
/**
 * Levenshtein Distance Algorithm (Optimized)
 */

function levenshteinDistanceOptimized(s1, s2) {
    if (s1.length < s2.length) {
        return levenshteinDistanceOptimized(s2, s1);
    }

    if (s2.length === 0) {
        return s1.length;
    }

    let previousRow = Array.from({ length: s2.length + 1 }, (_, i) => i);

    for (let i = 0; i < s1.length; i++) {
        let currentRow = [i + 1];

        for (let j = 0; j < s2.length; j++) {
            let insertions = previousRow[j + 1] + 1;
            let deletions = currentRow[j] + 1;
            let substitutions = previousRow[j] + (s1[i] !== s2[j] ? 1 : 0);
            currentRow.push(Math.min(insertions, deletions, substitutions));
        }
        previousRow = currentRow;
    }

    return previousRow[previousRow.length - 1];
}

// Driver code
const s1 = "kitten";
const s2 = "sitting";
console.log(`Distance between ${s1} and ${s2} is: ${levenshteinDistanceOptimized(s1, s2)}`);

File: strings/levenshtein_distance.rs
/**
 * Levenshtein Distance Algorithm (Optimized) in Rust
 */

use std::cmp::min;

fn levenshtein_distance_optimized(s1: &str, s2: &str) -> usize {
    let mut s1_chars: Vec<char> = s1.chars().collect();
    let mut s2_chars: Vec<char> = s2.chars().collect();

    if s1_chars.len() < s2_chars.len() {
        return levenshtein_distance_optimized(s2, s1);
    }

    if s2_chars.len() == 0 {
        return s1_chars.len();
    }

    let mut previous_row: Vec<usize> = (0..=s2_chars.len()).collect();

    for (i, c1) in s1_chars.iter().enumerate() {
        let mut current_row = vec![i + 1];

        for (j, c2) in s2_chars.iter().enumerate() {
            let insertions = previous_row[j + 1] + 1;
            let deletions = current_row[j] + 1;
            let substitutions = previous_row[j] + if c1 != c2 { 1 } else { 0 };
            current_row.push(min(insertions, min(deletions, substitutions)));
        }
        previous_row = current_row;
    }

    *previous_row.last().unwrap()
}

fn main() {
    let s1 = "kitten";
    let s2 = "sitting";
    println!("Distance between {} and {} is: {}", s1, s2, levenshtein_distance_optimized(s1, s2));
}
